clients/drcachesim/tests/record_filter_unit_tests.cpp

```diff
@@ -396,7 +410,141 @@ test_cache_and_type_filter()
+        { { TRACE_TYPE_INSTR, 2, { 0x1236 } }, true },
+        { { TRACE_TYPE_ENCODING, 2, { 0xf00d } }, false },
+        { { TRACE_TYPE_INSTR, 2, { 01237 } }, true },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CHUNK_FOOTER, { 1 } }, true },
+        { { TRACE_TYPE_FOOTER, 0, { 0xa2 } }, true }
+    };
+
+    // Test that the ordinal marker is updated on removing records.
+    auto stream = std::unique_ptr<local_stream_t>(new local_stream_t());
+    std::vector<std::unique_ptr<record_filter_func_t>> filters;
+    auto filter =
+        std::unique_ptr<record_filter_func_t>(new dynamorio::drmemtrace::type_filter_t(
+            { TRACE_TYPE_ENCODING },
+            { TRACE_MARKER_TYPE_FUNC_ID, TRACE_MARKER_TYPE_FUNC_RETADDR,
+              TRACE_MARKER_TYPE_FUNC_ARG }));
+    if (!filter->get_error_string().empty()) {
+        fprintf(stderr, "Couldn't construct a type_filter %s",
+                filter->get_error_string().c_str());
+        return false;
+    }
+    filters.push_back(std::move(filter));
+    auto record_filter = std::unique_ptr<test_record_filter_t>(
+        new test_record_filter_t(std::move(filters), 0, /*write_archive=*/true));
+    void *shard_data =
+        record_filter->parallel_shard_init_stream(0, nullptr, stream.get());
+    if (!*record_filter) {
+        fprintf(stderr, "Filtering init failed\n");
+        return false;
+    }
+    for (int i = 0; i < static_cast<int>(entries.size()); ++i) {
+        // We need to emulate the stream for the tool.
+        if (entries[i].entry.type == TRACE_TYPE_MARKER &&
+            entries[i].entry.size == TRACE_MARKER_TYPE_TIMESTAMP)
+            stream->set_last_timestamp(entries[i].entry.addr);
+        if (!record_filter->parallel_shard_memref(shard_data, entries[i].entry)) {
+            fprintf(stderr, "Filtering failed: %s\n",
+                    record_filter->parallel_shard_error(shard_data).c_str());
+            return false;
+        }
+    }
+    if (!record_filter->parallel_shard_exit(shard_data) || !*record_filter) {
+        fprintf(stderr, "Filtering exit failed\n");
+        return false;
+    }
+    // Check filtered output entries.
+    std::vector<trace_entry_t> filtered = record_filter->get_output_entries();
+    int j = 0;
+    for (int i = 0; i < static_cast<int>(entries.size()); ++i) {
+        if (!entries[i].remain)
+            continue;
+        if (j >= static_cast<int>(filtered.size())) {
+            fprintf(stderr, "Too few entries in filtered output. Expected: ");
+            print_entry(entries[i].entry);
+            fprintf(stderr, "\n");
+            return false;
+        }
+        if (entries[i].entry.type == TRACE_TYPE_MARKER &&
+            entries[i].entry.size == TRACE_MARKER_TYPE_RECORD_ORDINAL) {
+            if (filtered[j].type != entries[i].entry.type ||
+                filtered[j].size != entries[i].entry.size ||
+                filtered[j].addr != entries[i].entry.addr - REMOVED_VISIBLE_COUNT) {
```

Comment:
The other test adds entries for both the input and output stream in the std::vector<struct test_case>, and uses a bool to determine whether or not to use the entry as input to the tool. IMHO it makes it easier to read the expected output (replacement entries are right next to the input entries), and also avoids all this special casing. Could optionally use that strategy in this test too.

Reply:
When I first looked at it, it looked like I couldn't re-use much.  But on a second look: I was able to make a helper and share all the same checking code, changing the new test to use the combined form.  Done.